src/bitcoin-miner.cpp

```diff
@@ -0,0 +1,723 @@
+    evhttp_add_header(output_headers, "Connection", "close");
+    evhttp_add_header(
+        output_headers, "Authorization", (std::string("Basic ") + EncodeBase64(strRPCUserColonPass)).c_str());
+
+    // Attach request data
+    std::string strRequest = JSONRPCRequest(strMethod, params, 1);
+    struct evbuffer *output_buffer = evhttp_request_get_output_buffer(req);
+    assert(output_buffer);
+    evbuffer_add(output_buffer, strRequest.data(), strRequest.size());
+
+    int r = evhttp_make_request(evcon, req, EVHTTP_REQ_POST, "/");
+    if (r != 0)
+    {
+        evhttp_connection_free(evcon);
+        event_base_free(base);
+        throw CConnectionFailed("send http request failed");
+    }
+
+    event_base_dispatch(base);
+    evhttp_connection_free(evcon);
+    event_base_free(base);
+
+    if (response.status == 0)
+        throw CConnectionFailed("couldn't connect to server");
+    else if (response.status == HTTP_UNAUTHORIZED)
+        throw runtime_error("incorrect rpcuser or rpcpassword (authorization failed)");
+    else if (response.status >= 400 && response.status != HTTP_BAD_REQUEST && response.status != HTTP_NOT_FOUND &&
+             response.status != HTTP_INTERNAL_SERVER_ERROR)
+        throw runtime_error(strprintf("server returned HTTP error %d", response.status));
+    else if (response.body.empty())
+        throw runtime_error("no response from server");
+
+    // Parse reply
+    UniValue valReply(UniValue::VSTR);
+    if (!valReply.read(response.body))
+        throw runtime_error("couldn't parse reply from server");
+    const UniValue &reply = valReply.get_obj();
+    if (reply.empty())
+        throw runtime_error("expected reply to have result, error and id properties");
+
+    return reply;
+}
+
+
+static CBlockHeader CpuMinerJsonToHeader(const UniValue &params)
+{
+    // Does not set hashMerkleRoot (Does not exist in Mining-Candidate params).
+    CBlockHeader blockheader;
+
+    // nVersion
+    blockheader.nVersion = params["version"].get_int();
+
+    // hashPrevBlock
+    string tmpstr = params["prevhash"].get_str();
+    std::vector<unsigned char> vec = ParseHex(tmpstr);
+    std::reverse(vec.begin(), vec.end()); // sent reversed
+    blockheader.hashPrevBlock = uint256(vec);
+
+    // nTime:
+    blockheader.nTime = params["time"].get_int();
+
+    // nBits
+    {
+        std::stringstream ss;
+        ss << std::hex << params["nBits"].get_str();
+        ss >> blockheader.nBits;
+    }
+
+    return blockheader;
+}
+
+
+static void CalculateNextMerkleRoot(uint256 &merkle_root, uint256 &merkle_branch)
+{
+    // Append a branch to the root. Double SHA256 the whole thing:
+    uint256 hash;
+    CHash256()
+        .Write(merkle_root.begin(), merkle_root.size())
+        .Write(merkle_branch.begin(), merkle_branch.size())
+        .Finalize(hash.begin());
+    merkle_root = hash;
+}
+
+static uint256 CalculateMerkleRoot(uint256 &coinbase_hash, std::vector<uint256> merklebranches)
+{
```

Comment:
-std::vector<uint256> merklebranches
+std::vector<uint256> &merklebranches

Reply:
Thanks I missed that.